csrc/preseg_passes/pre_segmenter.cpp

```diff
@@ -22,16 +23,35 @@ namespace nvfuser::preseg_passes {
 /*static*/ void PreSegmenter::runPass(Fusion* fusion) {
   FUSER_PERF_SCOPE("PreSegmenter::runPass");
 
-  // Replace TensorViews with zero extent. Outputs and inputs may still be empty
-  OptimizationPass<RemoveEmptyPass>::runPass(fusion);
-  // removes consecutive cast operations
-  OptimizationPass<ConsecutiveCastPass>::runPass(fusion);
-  OptimizationPass<AddAxiomsPass>::runPass(fusion);
-  OptimizationPass<MoveSplitCatPass>::runPass(fusion);
-  OptimizationPass<MarkAliasesPreparePass>::runPass(fusion);
-  OptimizationPass<ExactMappedExtentSubstitutionPass>::runPass(fusion);
-  OptimizationPass<AllocationDomainPass>::runPass(fusion);
-  OptimizationPass<RemoveBcastSqueeze>::runPass(fusion);
+  // Put passes into a vector to run them in order
+  using PassFunction = std::function<void(Fusion*)>;
+  using TaggedPass = std::pair<PassFunction, std::string>;
+  std::vector<TaggedPass> passes = {
+      {OptimizationPass<RemoveEmptyPass>::runPass, "RemoveEmptyPass"},
+      {OptimizationPass<ConsecutiveCastPass>::runPass, "ConsecutiveCastPass"},
+      {OptimizationPass<AddAxiomsPass>::runPass, "AddAxiomsPass"},
+      {OptimizationPass<MoveSplitCatPass>::runPass, "MoveSplitCatPass"},
+      {OptimizationPass<MarkAliasesPreparePass>::runPass,
+       "MarkAliasesPreparePass"},
+      {OptimizationPass<ExactMappedExtentSubstitutionPass>::runPass,
+       "ExactMappedExtentSubstitutionPass"},
+      {OptimizationPass<AllocationDomainPass>::runPass, "AllocationDomainPass"},
+      {OptimizationPass<RemoveBcastSqueeze>::runPass, "RemoveBcastSqueeze"}};
+
+  bool is_log_enabled =
+      isDebugDumpEnabled(DebugDumpOption::PreSegmenterLogging);
+  if (is_log_enabled) {
+    debug() << "Fusion before PreSegmenter:" << std::endl;
+    fusion->printMath();
+  }
```

Comment:
Sorry, I didn't put too much thought on where exactly to log. 

How about moving this to after https://github.com/NVIDIA/Fuser/blob/c2dbfe4cf76fdce436d2899bf32fac0b2b1a0778/csrc/preseg_passes/optimization_pass.h#L53? The class OptimizationPass was created to handle common logics like this. Other examples include checking cycles and skipping disabled passes.

Reply:
That probably also requires to embed the tag/name in the pass. For example, each preseg pass will have a method `name()` that returns a string.

Reply:
moved to after `DerivedClass::runPass`, added `name()` to each pass.
When run with `NVFUSER_DUMP=pre_segmenter_logging`, a typical output is
```
[ RUN      ] NVFuserTest.FusionRemoveEmptyMatmul_CUDA
Fusion before PreSegmenter optimizations:
Inputs:
  T0_g[ iS0{16}, iS1{0} ], __half
  T1_g[ iS2{0}, iS3{8} ], __half
Outputs:
  T5_g[ iS13{16}, iS14{8}, rS15{0} ], float

%kernel_math {
T2_l[ iS4{16}, bS5{1}, iS6{0} ]
   = broadcast( T0_g[ iS0{16}, iS1{0} ] )
T3_l[ bS7{1}, iS8{0}, iS9{8} ]
   = broadcast( T1_g[ iS2{0}, iS3{8} ] )
T4_l[ bS10{1}, iS12{8}, iS11{0} ]
   = Set.Permute( T3_l[ bS7{1}, iS8{0}, iS9{8} ], cache_op=Streaming )
T5_g[ iS13{16}, iS14{8}, rS15{0} ]
   = mma(T2_l[ iS4{16}, bS5{1}, iS6{0} ],
         T4_l[ bS10{1}, iS12{8}, iS11{0} ])
}

========================================
Fusion after pass: RemoveEmptyPass
Inputs:
  T0_g[ iS0{16}, iS1{0} ], __half
  T1_g[ iS2{0}, iS3{8} ], __half
Outputs:
  T6_g[ iS16{16}, iS17{8} ], float

%kernel_math {
T6_g[ iS16{16}, iS17{8} ]
   = full({16, 8}, float(0));
}

========================================
Fusion after pass: ConsecutiveCastPass
Inputs:
  T0_g[ iS0{16}, iS1{0} ], __half
  T1_g[ iS2{0}, iS3{8} ], __half
Outputs:
  T6_g[ iS16{16}, iS17{8} ], float

%kernel_math {
T6_g[ iS16{16}, iS17{8} ]
   = full({16, 8}, float(0));
}
```